### PR TITLE
Voting procedure map

### LINKED ISSUE #3542: Convert `votingProceduresTxBodyL` to a `Map`
Currently `votingProceduresTxBodyL` is defined as a sequence, which is problematic because it allows duplicate proposals.

Duplicate proposals don't make sense in general, but in the same transaction they will even cause issues when identifying a script with a `ScriptPurpose`.

For this reasons 
```haskell
  , ctbrVotingProcedures :: !(StrictSeq (VotingProcedure era))
```
must be switched to:
```haskell
  , ctbrVotingProcedures :: !(Map (VotingProcedureId (EraCrypto era)) (VotingProcedure era))
...

data VotingProcedureId c = VotingProcedureId
  { vpiGovActionId :: !(GovActionid c)
  , vpiVoter :: !(Voter c)
  }
```

This means that the `VotingProcedure` will loose the above two fields:
```haskell
data VotingProcedure era = VotingProcedure
  { vProcVote :: !Vote
  , vProcAnchor :: !(StrictMaybe (Anchor (EraCrypto era)))
  }
```

### PR DESCRIPTION
**Pull request by lehins**
# Description

Fixes #3542 

The fix is slightly different than in the ticket. The `votingProceduresTxBodyL` field is switched to a nested Map that is enforced to not have any empty Maps on the second level by the decoder. Arbitrary instance also reflects this invariant.

This approach reduced transaction size significantly when there many votes for the same Voter, which is expected to be a common case. Also simplifies rules in logic a bit.

Also bumps up versions for `small-steps-test` and `cardano-ledger-shelley-test`, since those provide support for GHC-9.6, but haven't been released yet.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,13 @@
    * `TALLY` -> `GOV`
    * `VDEL` -> `GOVCERT`
 * Make `Anchor` required in `ProposalProcedure`.
+* Add `VotingProcedures` type
+* Remove `vProcGovActionId` and `vProcVoter` from `VotingProcedure`
+* Change the type of `votingProceduresL` to return `VotingProcedures`, which is a nested Map
+  instead of a sequence, as before.
+* Change `GovernanceActionDoesNotExist` to `GovernanceActionsDoNotExist`, which now
+  reports all actions as a set, rather than one action per each individual failure.
+* Type of `gpVotingProcedures` in `GovernanceProcedures` was aslo changed to `GovernanceProcedures`
```
Comment (Soupstraw) [inline eras/conway/impl/CHANGELOG.md:42]:
```suggestion
* Type of `gpVotingProcedures` in `GovernanceProcedures` was also changed to `GovernanceProcedures`
```

File: eras/conway/test-suite/cddl-files/conway.cddl [hunk 2]
```diff
@@ -67,15 +67,15 @@ transaction_body =
   , ? 16 : transaction_output       ; collateral return
   , ? 17 : coin                     ; total collateral
   , ? 18 : set<transaction_input>   ; reference inputs
-  , ? 19 : [* voting_procedure]     ; New; Voting procedures
+  , ? 19 : voting_procedures        ; New; Voting procedures
   , ? 20 : [* proposal_procedure]   ; New; Proposal procedures
   }
 
+voting_procedures = { + voter => { + governance_action_id => voting_procedure } }
```
Comment (Soupstraw) [inline eras/conway/test-suite/cddl-files/conway.cddl:74]: It looks to me like the actual decoder would happily decode an empty outer map, so the CDDL might be too strict here
```suggestion
voting_procedures = { voter => { + governance_action_id => voting_procedure } }
```

Comment (lehins) [reply eras/conway/test-suite/cddl-files/conway.cddl:74]: No, empty Map in Haskell means that the field is missing in CBOR.

See decoder for the `mint` field. It is the exact same approach.

**Review by Soupstraw — APPROVED** [review]
LGTM

**Review by aniketd — APPROVED** [review]
LGTM! :raised_hands:

**Pull request merged**